arch/eventloop_lws_http.c

```diff
@@ -0,0 +1,259 @@
+};
+
+static int
+callback_http(struct lws *wsi, enum lws_callback_reasons reason,
+              void *user, void *in, size_t len) {
+    switch (reason) {
+
+    case LWS_CALLBACK_ESTABLISHED_CLIENT_HTTP:
+        lwsl_user("LWS_CALLBACK_ESTABLISHED_CLIENT_HTTP: resp %u\n",
+                  lws_http_client_http_response(wsi));
+        break;
+
+        /* because we are protocols[0] ... */
+    case LWS_CALLBACK_CLIENT_CONNECTION_ERROR:
+        lwsl_err("CLIENT_CONNECTION_ERROR: %s\n",
+                 in ? (char *)in : "(null)");
+        break;
+
+        /* chunks of chunked content, with header removed */
+    case LWS_CALLBACK_RECEIVE_CLIENT_HTTP_READ:
+        lwsl_user("RECEIVE_CLIENT_HTTP_READ: read %d\n", (int)len);
+        lwsl_hexdump_info(in, len);
+        UA_LOG_INFO(UA_Log_Stdout, UA_LOGCATEGORY_NETWORK,
+                    "RECEIVE_CLIENT_HTTP_READ: %s", (char*)in);
+        return 0; /* don't passthru */
+
+        /* uninterpreted http content */
+    case LWS_CALLBACK_RECEIVE_CLIENT_HTTP:
+    {
+        char buffer[1024 + LWS_PRE];
+        char *px = buffer + LWS_PRE;
+        int lenx = sizeof(buffer) - LWS_PRE;
+
+        if (lws_http_client_read(wsi, &px, &lenx) < 0)
+            return -1;
+    }
+        return 0; /* don't passthru */
+
+    case LWS_CALLBACK_COMPLETED_CLIENT_HTTP:
+        lwsl_user("LWS_CALLBACK_COMPLETED_CLIENT_HTTP %s\n",
+                  lws_wsi_tag(wsi));
+        break;
+
+    case LWS_CALLBACK_CLOSED_CLIENT_HTTP:
+        lwsl_info("%s: closed: %s\n", __func__, lws_wsi_tag(wsi));
+        break;
+
+    default:
+        break;
+    }
+    return 0;
+}
+/* usually defined as const */
+static struct lws_protocols protocols[] = {
+    { "http", callback_http, 0, 0, 0, NULL, 0},
+    LWS_PROTOCOL_LIST_TERM
+};
+
+struct HTTPConnectionManager {
+    UA_ConnectionManager cm;
+    struct lws_context *context;
+};
+
+static UA_StatusCode
+HTTP_eventSourceStart(UA_ConnectionManager *cm) {
+    UA_EventLoop *el = cm->eventSource.eventLoop;
+    if(!el)
+        return UA_STATUSCODE_BADINTERNALERROR;
+
+    /* Check the state */
+    if(cm->eventSource.state != UA_EVENTSOURCESTATE_STOPPED) {
+        UA_LOG_ERROR(el->logger, UA_LOGCATEGORY_NETWORK,
+                     "HTTP\t| To start the ConnectionManager, it has to be "
+                     "registered in an EventLoop and not started yet");
+        return UA_STATUSCODE_BADINTERNALERROR;
+    }
+
+    cm->eventSource.state = UA_EVENTSOURCESTATE_STARTED;
+    return UA_STATUSCODE_GOOD;
+}
+
+static void
+HTTP_eventSourceStop(UA_ConnectionManager *cm) {
+    if(cm->eventSource.state == UA_EVENTSOURCESTATE_STOPPING ||
+       cm->eventSource.state == UA_EVENTSOURCESTATE_STOPPED)
+        return;
+
+    UA_LOG_INFO(cm->eventSource.eventLoop->logger, UA_LOGCATEGORY_NETWORK,
+                "HTTP\t| Shutting down the ConnectionManager");
+
+    cm->eventSource.state = UA_EVENTSOURCESTATE_STOPPING;
```

Comment:
At this point you should forcefully close the sockets.
That also means you need to keep a list of the open sockets.
See how _stop is done for TCP and UDP.